### ISSUE TITLE
Exporting FS should make sure it exists

### ISSUE THREAD

**Issue by saschanaz** [emscripten-core/emscripten#4515]
Adding `Module.FS=FS;` line before `return Module` line fixes this bug, but is this intentional?

**Comment by kripken** [comment]
I don't think `MODULARIZE` affects this - we just don't export (place a reference to it on `Module`) the FS object by default.


If you do want to export it, you can do `-s 'EXTRA_EXPORTED_RUNTIME_METHODS=["FS"]'` (which emits a line like the manual one you wrote).

**Comment by SaschaNaz** [comment]
Works well, thanks :D I should have searched more.

**Comment by SaschaNaz** [comment]
I thought it worked but the browser cache made me though that. `EXTRA_EXPORTED_RUNTIME_METHODS=["FS"]` does not add any line for me, and I skimmed some code and found that there is no `maybeExport` call for "FS".

**Comment by kripken** [comment]
It works here, what version are you on?

**Comment by SaschaNaz** [comment]
I'm on 1.35.0. I tried inserting `throw new Error(EXTRA_EXPORTED_RUNTIME_METHODS);` on `maybeExport` and it correctly shows `FS`, but I cannot understand why it does not write my line.

BTW, can I use `console.log` to debug on `module.js`? It does not show anything on the console.

**Comment by SaschaNaz** [comment]
`if ("FS" === name) throw new Error("FS!");` line on `maybeExport` never throws. Hmm.

**Comment by SaschaNaz** [comment]
Exporting FS is enabled on Nov 2015 and this is why I'm in trouble!  https://github.com/kripken/emscripten/commit/48c8a37a687dce95cb8cfe21bc7055036da8c769

**Comment by kripken** [comment]
Almost a year ago, I guess that's why I didn't remember fixing this ;)

**Comment by xiyanggudao** [comment]
I see that it allow easy exporting FS in this commit:
https://github.com/kripken/emscripten/commit/48c8a37a687dce95cb8cfe21bc7055036da8c769
I upgrade emscripten to 1.37.0, but still do not kown how to export FS. I try EXTRA_EXPORTED_RUNTIME_METHODS=["FS"], but get warning FS is undefined, it seems like that EXTRA_EXPORTED_RUNTIME_METHODS only support functions while FS is an object.
I try add a line "Module['FS'] = Module.FS = FS;" to postamble.js and find FS is exported successfully, is it the right way to export FS ?

**Comment by kripken** [comment]
It seems to work for me on incoming,

````
$ ./emcc tests/hello_world.c -o a.out.js
$ ./emcc tests/hello_world.c -s 'EXTRA_EXPORTED_RUNTIME_METHODS=["FS"]' -o b.out.js && diff a.out.js b.out.js9050c9050
< 
---
> Module["FS"] = FS;
````

so it does export `FS` properly. `FS` also needs to exist, though, and by default we strip it out. Using `-s FORCE_FILESYSTEM=1` can fix that. We should probably automatically do that when it is exported, though, so I think that's a bug - worth fixing, hopefully someone has the time.

**Comment by dphil** [comment]
Glad I found this. I was getting pretty confused why my modularized object was showing a handful of FS functions (such as "FS_createDataFile" and a few others) attached to it, but not the bulk of them, nor any attached or global "FS" object that I kept seeing in other documentation/examples. I noticed "FS" appeared in the global namespace if I built without MODULARIZE, so I was going to have to settle for a non-modularized build. I didn't realize I could just use 'EXTRA_EXPORTED_RUNTIME_METHODS=["FS"]'. That does the trick.
